lib/Dialect/SV/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,298 @@
+  auto *ctx = builder.getContext();
+  // Update the types of the source module ports based on evaluating any
+  // parametric in/output ports.
+  auto ports = source.getPorts();
+  for (auto &in : llvm::enumerate(source.getFunctionType().getInputs())) {
+    auto resType =
+        evaluateParametricType(source.getLoc(), parameters, in.value());
+    if (failed(resType))
+      return failure();
+    ports.inputs[in.index()].type = resType.getValue();
+  }
+  for (auto &out : llvm::enumerate(source.getFunctionType().getResults())) {
+    auto resType =
+        evaluateParametricType(source.getLoc(), parameters, out.value());
+    if (failed(resType))
+      return failure();
+    ports.outputs[out.index()].type = resType.getValue();
+  }
+
+  // Create the specialized module using the evaluated port info.
+  target = builder.create<HWModuleOp>(
+      source.getLoc(),
+      StringAttr::get(ctx, generateModuleName(source, parameters)), ports);
+
+  // Erase the default created hw.output op - we'll copy the correct operation
+  // during body elaboration.
+  (*target.getOps<hw::OutputOp>().begin()).erase();
+
+  // Clone body of the source into the target.
+  BlockAndValueMapping mapper;
+  for (auto &&[src, dst] :
+       llvm::zip(source.getArguments(), target.getArguments()))
+    mapper.map(src, dst);
+  builder.setInsertionPointToStart(target.getBodyBlock());
+  for (auto &op : source.getOps())
+    builder.clone(op, mapper);
+
+  // We've now created a separate copy of the source module with a rewritten
+  // top-level interface. Next, we enter the module to convert parametric types
+  // within operations.
+  RewritePatternSet patterns(ctx);
+  TypeConverter t;
+  populateTypeConversion(target.getLoc(), t, parameters);
+  patterns.add<EliminateParamValueOpPattern>(ctx, parameters);
+  patterns.add<NarrowArrayGetIndexPattern>(ctx);
+  patterns.add<ParametricTypeConversionPattern>(ctx, t, parameters);
+  ConversionTarget convTarget(*ctx);
+  convTarget.addLegalOp<hw::HWModuleOp>();
+  convTarget.addIllegalOp<hw::ParamValueOp>();
+
+  // Generic legalization of converted operations.
+  convTarget.addDynamicallyLegalDialect<comb::CombDialect, hw::HWDialect,
+                                        sv::SVDialect>(
+      [](Operation *op) { return !isParametricOp(op); });
+
+  return applyPartialConversion(target, convTarget, std::move(patterns));
+}
+
+void HWSpecializePass::runOnOperation() {
+  ModuleOp module = getOperation();
+
+  // Record unique module parameters and references to these.
+  llvm::DenseMap<hw::HWModuleOp, llvm::SetVector<ArrayAttr>>
+      uniqueModuleParameters;
+  llvm::DenseMap<hw::HWModuleOp,
+                 llvm::DenseMap<ArrayAttr, llvm::SmallVector<hw::InstanceOp>>>
+      parametersUsers;
+  for (auto hwModule : module.getOps<hw::HWModuleOp>()) {
+    for (auto instanceOp : hwModule.getOps<hw::InstanceOp>()) {
+      auto *targetOp = instanceOp.getReferencedModule();
```

Comment:
IIRC `getReferencedModule()` is a linear search.

Reply:
Modified to use the symbol cache.